### PR TITLE
Panwang/shim

### PR DESCRIPTION
**Pull request by pan-wang**
organize the code so that: 1) remove storing shim configuration on IIS module context and references. 
2) remove application_info_key class and use the application ID string as hashtable key
3) adding a failure flag in application info to avoid flooding event log in case of configuration error

### CODE REVIEW

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 1]
```diff
@@ -89,57 +76,47 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
         }
-
-        //
-        // first application will decide which hosting model allowed by this process
-        //
-        if (m_hostingModel == HOSTING_UNKNOWN)
+        else
         {
-            m_hostingModel = pConfig->QueryHostingModel();
+            if ((*ppApplicationInfo)->QueryConfig()->QueryHostingModel() == HOSTING_OUT_PROCESS &&
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: Store (*ppApplicationInfo)->QueryConfig() into a variable.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 2]
```diff
@@ -149,54 +126,39 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
-                ASPNETCORE_EVENT_MIXED_HOSTING_MODEL_ERROR_MSG,
-                pszApplicationId,
-                pConfig->QueryHostingModel())))
-            {
-                UTILITY::LogEvent(g_hEventLog,
-                    EVENTLOG_ERROR_TYPE,
-                    ASPNETCORE_EVENT_MIXED_HOSTING_MODEL_ERROR,
-                    strEventMsg.QueryStr());
-            }
+            UTILITY::LogEvent(g_hEventLog,
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: LogEventF?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 3]
```diff
@@ -89,57 +76,47 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
+            if ((*ppApplicationInfo)->QueryConfig()->QueryHostingModel() == HOSTING_OUT_PROCESS &&
+                (*ppApplicationInfo)->QueryConfig()->QueryHostingModel() == m_hostingModel)
+            {
+                (*ppApplicationInfo)->UpdateAllowStartStatus(TRUE);
+            }
+            else
+            {
+                if ((*ppApplicationInfo)->QueryConfig()->QueryHostingModel() != m_hostingModel)
+                {
+                    fMixedHostingModelError = TRUE;
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: no, we don't want to call report event inside an exclusive lock which blocks the activation of other applications

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 4]
```diff
@@ -89,57 +76,47 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
-        hr = pApplicationInfo->Initialize(pConfig, m_pFileWatcher);
+        hr = m_pApplicationInfoHash->InsertRecord(pApplicationInfo);
         if (FAILED(hr))
         {
             goto Finished;
         }
 
-        hr = m_pApplicationInfoHash->InsertRecord( pApplicationInfo );
-        if (FAILED(hr))
+        *ppApplicationInfo = pApplicationInfo;
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: done

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 5]
```diff
@@ -89,57 +76,47 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
+        pApplicationInfo = NULL;
+
+        if(m_pApplicationInfoHash->Count() ==1)
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: done

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx (hunk 2 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: nice catch

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx (hunk 1 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: yep

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 6]
```diff
@@ -41,38 +41,51 @@ APPLICATION_INFO::~APPLICATION_INFO()
...
 
-    m_pConfiguration = pConfiguration;
+    // todo: make sure Initialize should be called only once
+    m_pServer = pServer;
+    m_pConfiguration = new ASPNETCORE_SHIM_CONFIG();
 
-    // reference the configuration instance to prevent it will be not release
-    // earlier in case of configuration change and shutdown
-    m_pConfiguration->ReferenceConfiguration();
+    if (m_pConfiguration == NULL)
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: I did not see try/catch and make_unique in shim. For now I just want to make the code consistent in shim.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 7]
```diff
@@ -89,56 +77,47 @@ APPLICATION_MANAGER::GetOrCreateApplicationInfo(
...
-        //
-        // first application will decide which hosting model allowed by this process
-        //
-        if (m_hostingModel == HOSTING_UNKNOWN)
+        else
         {
-            m_hostingModel = pConfig->QueryHostingModel();
+            if (hostingModel == HOSTING_OUT_PROCESS &&  hostingModel == m_hostingModel)
+            {
+                (*ppApplicationInfo)->UpdateAllowStartStatus(TRUE);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: This will AV. Use pApplicationInfo.

Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: nice catch

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 8]
```diff
@@ -608,6 +608,15 @@ APPLICATION_INFO::RecycleApplication()
...
                 NULL);      // receive thread identifier
         }
+        else
+        {
+            if (m_pConfiguration->QueryHostingModel() == HOSTING_IN_PROCESS)
+            {
+                // In process application failed to start for whatever reason, need to recycle the work process
+                m_pServer->RecycleProcess(L"AspNetCore InProcess Recycle Process on Demand");
+            }
+
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: nit space.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: If you don't use make_unique  then you should delete m_pConfiguration if FILE_WATCHER_ENTRY fails to allocate.

[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: the destructor handles it.

**Pull request merged**